name.abuchen.portfolio/src/name/abuchen/portfolio/online/impl/LeewayQuoteFeed.java

```diff
@@ -0,0 +1,249 @@
+
+        if (!security.getPrices().isEmpty())
+            quoteStartDate = security.getPrices().get(security.getPrices().size() - 1).getDate();
+        else
+            quoteStartDate = LocalDate.of(2000, 01, 01);
+
+        return getHistoricalQuotes(security, collectRawResponse, quoteStartDate);
+    }
+
+    @Override
+    public QuoteFeedData previewHistoricalQuotes(Security security)
+    {
+        return getHistoricalQuotes(security, true, LocalDate.now().minusMonths(2));
+    }
+
+    @SuppressWarnings("unchecked")
+    public QuoteFeedData getHistoricalQuotes(Security security, boolean collectRawResponse, LocalDate quoteStartDate)
+    {
+        String securityTickerSymbol = trim(security.getTickerSymbol()).toUpperCase();
+
+        if (apiKey == null)
+        {
+            PortfolioLog.error(Messages.MsgErrorLeewayAPIKeyMissing);
+            return QuoteFeedData.withError(new IllegalArgumentException(Messages.MsgErrorLeewayAPIKeyMissing));
+        }
+
+        if (securityTickerSymbol == null || securityTickerSymbol.isEmpty())
+        {
+            PortfolioLog.error(MessageFormat.format(Messages.MsgMissingTickerSymbol, security.getName()));
+            return QuoteFeedData.withError(
+                            new IOException(MessageFormat.format(Messages.MsgMissingTickerSymbol, security.getName())));
+        }
+
+        QuoteFeedData data = new QuoteFeedData();
+
+        try
+        {
+            @SuppressWarnings("nls")
+            WebAccess webaccess = new WebAccess("api.leeway.tech", //
+                            "/api/v1/public/historicalquotes/" + securityTickerSymbol) //
+                                            .addParameter("apitoken", apiKey) //
+                                            .addParameter("from", quoteStartDate.toString());
+
+            // Lookup cached response for securityTickerSymbol
+            ResponseData response = cache.lookup(securityTickerSymbol);
+
+            // Check if cached response is null or if it has an earlier
+            // quoteStartDate than the requested quoteStartDate
+            if (response == null || response.quoteStartDate.isAfter(quoteStartDate))
+            {
+                JSONObject jsonObject = new JSONObject();
+                JSONArray jsonArray = (JSONArray) JSONValue.parse(webaccess.get());
+                jsonObject.put("prices", jsonArray); //$NON-NLS-1$
+
+                response = new ResponseData();
+                response.quoteStartDate = quoteStartDate;
+                response.json = jsonObject.toJSONString();
+
+                if (response.json != null)
+                    cache.put(securityTickerSymbol, response);
+            }
+
+            if (collectRawResponse)
+                data.addResponse(webaccess.getURL(), response.json);
+
+            JSONObject json = (JSONObject) JSONValue.parse(response.json);
+
+            JSONArray pricesJson = (JSONArray) json.get("prices"); //$NON-NLS-1$
```

Comment:
I do not understand the reason for this piece of code:
* First, I think it is misleading to the user to change the response. The response is an array, but we would display it as an object with the property "prices" that has the array. It is not what the server response.
* Second, the test "Response.json != null" does not make sense - it is always not null because we set it one line before
* Third, when parsing, you again extract the array out of the json object.
* Fourth, it looks like we parse the JSON twice - one in line 146, and then again in line 160.
Why not use the array directly?